pkg/templates/danglingservicemonitor/template.go

```diff
@@ -0,0 +1,70 @@
+package danglingservicemonitor
+
+import (
+	"fmt"
+
+	k8sMonitoring "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	"golang.stackrox.io/kube-linter/pkg/check"
+	"golang.stackrox.io/kube-linter/pkg/config"
+	"golang.stackrox.io/kube-linter/pkg/diagnostic"
+	"golang.stackrox.io/kube-linter/pkg/lintcontext"
+	"golang.stackrox.io/kube-linter/pkg/objectkinds"
+	"golang.stackrox.io/kube-linter/pkg/templates"
+	"golang.stackrox.io/kube-linter/pkg/templates/danglingservicemonitor/internal/params"
+	v1 "k8s.io/api/core/v1"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+func init() {
+	templates.Register(check.Template{
+		HumanName:   "Dangling Service Monitor",
+		Key:         "dangling-servicemonitor",
+		Description: "Flag service monitors which do not match any service",
+		SupportedObjectKinds: config.ObjectKindsDesc{
+			ObjectKinds: []string{objectkinds.ServiceMonitor},
+		},
+		Parameters:             params.ParamDescs,
+		ParseAndValidateParams: params.ParseAndValidate,
+		Instantiate: params.WrapInstantiateFunc(func(_ params.Params) (check.Func, error) {
+			return func(lintCtx lintcontext.LintContext, object lintcontext.Object) []diagnostic.Diagnostic {
+				serviceMonitor, ok := object.K8sObject.(*k8sMonitoring.ServiceMonitor)
+				if !ok {
+					return nil
+				}
+				nsSelector := serviceMonitor.Spec.NamespaceSelector.MatchNames
+				labelSelectors := serviceMonitor.Spec.Selector.MatchLabels
+				if len(labelSelectors) == 0 && len(nsSelector) == 0 {
+					return []diagnostic.Diagnostic{{
+						Message: "service monitor has no label or namespace selector specified",
+					}}
+				}
+				labelSelector, err := metaV1.LabelSelectorAsSelector(&metaV1.LabelSelector{MatchLabels: serviceMonitor.Spec.Selector.MatchLabels})
+				if err != nil {
+					return []diagnostic.Diagnostic{{
+						Message: fmt.Sprintf("service monitor has invalid label selector: %v", err),
+					}}
+				}
+				for _, obj := range lintCtx.Objects() {
+					services, found := obj.K8sObject.(*v1.Service)
+					if !found {
+						continue
+					}
+					if len(nsSelector) != 0 {
```

Comment:
This is not correct https://prometheus-operator.dev/docs/operator/api/#monitoring.coreos.com/v1.NamespaceSelector

```
// NamespaceSelector is a selector for selecting either all namespaces or a
// list of namespaces.
// If `any` is true, it takes precedence over `matchNames`.
// If `matchNames` is empty and `any` is false, it means that the objects are
// selected from the current namespace.
```

how about extracting selector code to a helper function?

Reply:
good idea, I'll work on fixing how we handle the namespace selector, moving it to a helper function, and configuring the selector check tomorrow.
Thanks for all the feedback!

Reply:
Corrected the logic and moved it to a helper fn in [116b187](https://github.com/stackrox/kube-linter/pull/547/commits/116b1870a1c0f7d5e15452db6758f4abbda16b59).

Next I'll look into configuring the selector check

Reply:
I've added the ability to disable the namespace/label selector checks- any feedback?